compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java

```diff
@@ -702,10 +667,100 @@ private static BIntersectionType defineImmutableUnionType(Location pos, Types ty
+        unionTypeNode.setBType(effectiveType);
+        BLangTypeDefinition typeDefinition = TypeDefBuilderHelper.addTypeDefinition(effectiveType,
+                                                                                    effectiveType.tsymbol,
+                                                                                    unionTypeNode, env);
+        typeDefinition.pos = pos;
+        return immutableType;
+    }
+
+    private static BIntersectionType defineImmutableBuiltInUnionType(Location pos, Types types, SymbolEnv env,
+                                                                     PackageID pkgId, BSymbol owner,
+                                                                     SymbolTable symTable,
+                                                                     BLangAnonymousModelHelper anonymousModelHelper,
+                                                                     Names names, Set<BType> unresolvedTypes,
+                                                                     BUnionType type, BType originalType) {
+        BTypeSymbol origBuiltInUnionTypeSymbol = type.tsymbol;
+
+        LinkedHashSet<BType> originalMemberList = type.getMemberTypes();
+        Optional<BIntersectionType> immutableTypeOptional = Types.getImmutableType(symTable, pkgId, type);
+        if (immutableTypeOptional.isPresent()) {
+            return immutableTypeOptional.get();
+        }
+
+        if (type.tag == TypeTags.JSON) {
+            defineImmutableJsonType(env, pkgId, owner, symTable, names, (BJSONType) type, originalType);
+        } else {
+            defineImmutableAnydataType(env, pkgId, owner, symTable, names, (BAnydataType) type, originalType);
+        }
+
+        return handleImmutableUnionType(pos, types, env, pkgId, owner, symTable, anonymousModelHelper, names,
+                                        unresolvedTypes, type, origBuiltInUnionTypeSymbol, originalMemberList);
+    }
+
+    private static void defineImmutableAnydataType(SymbolEnv env, PackageID pkgId, BSymbol owner, SymbolTable symTable,
+                                                   Names names, BAnydataType type, BType originalType) {
+        BTypeSymbol immutableAnydataTSymbol = getReadonlyTSymbol(names, type.tsymbol, env, pkgId, owner);
+
+        BAnydataType immutableAnydataType;
+        if (immutableAnydataTSymbol != null) {
+            immutableAnydataType =
+                    new BAnydataType(immutableAnydataTSymbol,
+                                     immutableAnydataTSymbol.name, type.flags | Flags.READONLY,
+                                     type.isNullable());
+            immutableAnydataTSymbol.type = immutableAnydataType;
+        } else {
+            immutableAnydataType =
+                    new BAnydataType(immutableAnydataTSymbol,
+                                     getImmutableTypeName(names, TypeKind.ANYDATA.typeName()),
```

Comment:
Why did you set the name here? Similar to the immutablejson, if the type symbol is null, why not set an empty string as the name?

Reply:
Would make it less accurate though? JSON doesn't require a name.

Reply:
Why does JSON not require a name, but AnyData does?

Reply:
Create a general issue to track this https://github.com/ballerina-platform/ballerina-lang/issues/39498